### ISSUE TITLE
Allow x=1 no-spaces variable asignment

### ISSUE THREAD

**Issue by eugenesvk** [nushell/nushell#13404]
### Related problem

It's pretty common to just do `let x=1` without any extra spaces to set values, but nushell is too strict
```
~> let a=2
Error: nu::parser::parse_mismatch

  × Parse mismatch during operation.
   ╭─[entry #8:1:5]
 1 │ let a=2
   ·     ─┬─
   ·      ╰── expected valid variable name
   ╰────
```

### Describe the solution you'd like

Relax rules and allow skipping spaces

### Describe alternatives you've considered

n/a

**Comment by fdncred** [comment]
Nushell has always been space sensitive. I'm not sure that will change.

**Comment by eugenesvk** [comment]
nushell is too young to be blocked by "always been"

**Comment by sholderbach** [comment]
While there are certainly some areas where we could relax the "just add a space" requirements (I haven't looked in depth at the assignment case, some of it is pretty unique syntax where you could relax some of it if you bend the lexing to your will...) we have a ton of things that would easily become ambiguous if you expect to be able to elide a space.
We went through a in my recollection horrible phase trying to permit math operators without space and that in combination also with variable access without `$` and this certainly had some huge other drawbacks. E.g. strong constraints on variable names. More quoting of words that could otherwise be bare, in general a much harder time getting both external commands with free parsing rules and nushells internal syntax to play nice.
Even if there may be some special cases where we can relax the whitespace requirements, there is some teaching incentive to make it a default for those case where it is not just "be nice to the lexing complexity" but otherwise it can't be disambiguated.

**Comment by KAAtheWiseGit** [comment]
Another problem is that even if space elision gets permitted in one place, there are still cases in which it's impossible to allow it.  `1+2` can't really evaluate to 3, because it might be a name of a binary.  So, it'll be very confusing to be able to omit spaces in one place, but having to add them in another.

On a similar note, that's also the reason I don't like the `ENV_VAR=1 command` syntax.  It's very Bash/POSIX sh and doesn't mesh well with the existing Nushell grammar rules.

**Comment by eugenesvk** [comment]
> `1+2` can't really evaluate to 3, because it might be a name of a binary

How often would you encounter that vs adding numbers? I'd take the need to quote more in the rare cases with binaries consisting of all numbers and pluses vs being unable to do basic math

> `ENV_VAR=1 command` syntax. It's very Bash/POSIX sh

`ENV_VAR = 1 command` doesn't work now, so it would continue not to work without spaces?

**Comment by eugenesvk** [comment]
> permit math operators without space and that in combination also with variable access without `$` and this certainly had some huge other drawbacks.

Well, one is totally not like the other:`$` is indeed a huge impact on common collision cases.

Huge issues wtih no-space arithmetic, on the other hand, is puzzling since it's so common, even in shells. Is there a list/discussion on what very common issues arise from that I could check?

> there is some teaching incentive to make

I think the good lesson that should be taught is "resolve ambiguity when required" rather than "think about it all the time"

**Comment by KAAtheWiseGit** [comment]
The way I see it, currently in Nushell the rule is to always put spaces between all operations.  In that way it's closer to how programming languages are usually written, and I personally prefer that.

The biggest problem I see with allowing space elision is inconsistency.  There are dozens of operators, and some of them must be space separated (i.e. `starts-with`).  So, an elision rule would probably be quite complex and non-trivial to remember.

So, I'd rather have more spaced out code with consistent rules, rather than a more dense code with more complex grammar rules.

**Comment by KAAtheWiseGit** [comment]
I also recall there being a separate discussion, about allowing braces in function definitions to start on a new line.  I'll try to find it, I think it'll have a lot of relevant points.

**Comment by eugenesvk** [comment]
> In that way it's closer to how programming languages are usually written, and I personally prefer that.

which usual programming languages ban `x=1` without spaces?

> The biggest problem I see with allowing space elision is inconsistency.

That's indeed the biggest problem that prompted this issue - no spaces is inconsistent with your experience everywhere else

> some of them must be space separated (i.e. `starts-with`).

Don't see the connection to `=`: `starts-with` consists of words, words have almost always required a separator (space/- etc (or at a bare minimum Capitalization))

**Comment by fdncred** [comment]
@eugenesvk You should feel free to submit a PR for the parser that makes nushell work the way you think it should. I'm not guaranteeing that said PR would be accepted but it could be if there are no down sides to it. I tend to think such a change would be incompatible with the rest of nushell and create loads of papercuts, but I could be wrong and am willing to be proven so.

**Comment by sholderbach** [comment]
As a result from the space-free math attempt we had to restrict variable names (and kept this):
```nushell
> let test-1 = "ERR"
Error: nu::parser::parse_mismatch

  × Parse mismatch during operation.
   ╭─[entry #1:1:5]
 1 │ let test-1 = "ERR"
   ·     ───┬──
   ·        ╰── expected valid variable name
   ╰────
> let test_1 = "OK"
```

Motivated by the possible confusion in:
```nushell
$test-1
```

On top of consistency with "how things are done in other languages" this lead us also to introduce a name translation for command flags:
```nushell
def foo [--bar-with-dash] {
  $bar_with_dash
}
```
---

One example where our current rules disfavor space-free operators:
Record keys/columns have the requirement to support arbitrary strings and we are relatively relaxed on the cell path quoting requirements.
```nushell
> let foo = {"test": 42, "test-1": 13}
> $foo.test-1
13
```
---
As @fdncred pointed out, feel free to flesh out your proposal as a PR, but there are many subtle places to consider for general ergonomics, so the burden of proof is on the ones making the change.

**Comment by NotTheDr01ds** [comment]
I love the Nushell parser; it almost feels magical to me.  And so I'm willing to live with things like space-separation and a few others.  I even like them now.

As @KAAtheWiseGit said, and I fully agree with ...

> So, I'd rather have more spaced out code with consistent rules, rather than a more dense code with more complex grammar rules.

That said, I do personally see this as "nice to have" assuming it didn't negatively affect overall ergonomics.

> nushell is too young to be blocked by "always been"

I agree, and meant to comment back when I saw this originally.

> I think the good lesson that should be taught is "resolve ambiguity when required" rather than "think about it all the time"

This is a completely fair point, IMHO, and one I've made before as well in other context. Another way to think about it is that the language/parser should aim to make the *most common use-case* the easiest, and the exceptions can be a little more difficult.  *If* (big if) the parser could handle this without negatively affecting other *common* cases, I'd be okay with it.

For instance, I'd be happy giving up that current `$foo.test-1` syntax and requiring `$foo."test-1"` (which already works).  The readability of the former is *not great* (again, IMHO).  Someone reading this at first glance would likely see it as a subtraction.  Only the most experienced Nushell user would see `test-1` as an index/key.  That said, it's purely *my* opinion, and others might see bare-word keys as a big ergonomic advantage.

> @fdncred a change would be incompatible with the rest of nushell and create loads of papercuts, but I could be wrong and am willing to be proven so.

Agreed, there are so many places this change could hit that might (just guessing) impact ergonomics of *common* scenarios.  I'd love to be wrong about this as well.

> Is there a list/discussion on what very common issues arise from that I could check?

Probably not.  It's likely going to require someone to make the change and see what tests fail as a result.

I'm not sure what your Rust skills are; I know that many of us are users of Nushell and not able to contribute changes like this as fully as we'd like.  I've started learning Rust in order to make enhancements and fixes that I can, but (a) the parser is beyond my skill level by far, and (b) I'm okay with the current way myself so I'm unlikely to ever take this up personally.

It will basically require someone who feels strongly enough about it, and has the appropriate skill level, to make the change and report back.

**Comment by eugenesvk** [comment]
> As a result from the space-free math attempt we had to restrict variable names (and kept this):

So you don't even get the benefit of banning space-free math if the restrictions are kept?

> Motivated by the possible confusion in:
> 
> ```nushell
> $test-1
> ```

But this exact confusion remains due to `$foo.test-1` being allowed

> One example where our current rules disfavor space-free operators: Record keys/columns have the requirement to support arbitrary strings and we are relatively relaxed on the cell path quoting requirements.
> 
> ```nushell
> > let foo = {"test": 42, "test-1": 13}
> > $foo.test-1
> 13
> ```

But then `=` still fails despite the space-free math banned
```nushell
> let foo = {"t":0, "t-1":1, "t+2":2, "t=3":3}
> $foo.t=3
Error: nu::parser::unknown_command
```

> As @fdncred pointed out, feel free to flesh out your proposal as a PR, but there are many subtle places to consider for general ergonomics, so the burden of proof is on the ones making the change.

But nu is the one making the change vs the more "general" occurence in other language/shell syntaxex, so apparently the burden has been met by whatever proof there already was. And given the strong sentiment against the change expressed here from some of the core project members it doesn't make sense to invest much time in testing a PR
